File: services/base/workflow-api/docker/files/app/api/v1/routers/workflows.py
import logging
import uuid
from typing import List, Optional

from app import schemas
from app.api.v1.services import workflow_service as service
from app.dependencies import get_async_db
from fastapi import APIRouter, Depends, Response
from sqlalchemy.ext.asyncio import AsyncSession

logger = logging.getLogger(__name__)
router = APIRouter()

# workflows


@router.get("/workflows", response_model=List[schemas.Workflow])
async def get_workflows(
    skip: int = 0,
    limit: int = 100,
    order_by: Optional[str] = None,
    order: Optional[str] = "desc",
    id: Optional[uuid.UUID] = None,
    title: Optional[str] = None,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.get_workflows(db, skip, limit, order_by, order, id, title)


@router.post("/workflows", response_model=schemas.Workflow, status_code=201)
async def create_workflow(
    workflow: schemas.WorkflowCreate,
    response: Response,
    db: AsyncSession = Depends(get_async_db),
):
    workflow_res = await service.create_workflow(db, workflow)
    response.headers["Location"] = f"/workflows/{workflow_res.id}"
    return workflow_res


@router.get("/workflows/{workflow_id}", response_model=schemas.Workflow)
async def get_workflow_by_id(
    workflow_id: uuid.UUID,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.get_workflow_by_id(db, workflow_id)


@router.patch("/workflows/{workflow_id}", response_model=schemas.Workflow)
async def update_workflow(
    workflow_id: uuid.UUID,
    update: schemas.WorkflowUpdate,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.update_workflow(db, workflow_id, update)


@router.delete("/workflows/{workflow_id}", status_code=204)
async def delete_workflow(
    workflow_id: uuid.UUID,
    db: AsyncSession = Depends(get_async_db),
):
    await service.delete_workflow(db, workflow_id)


# revisions


@router.get(
    "/workflows/{workflow_id}/revisions",
    response_model=List[schemas.WorkflowRevision],
)
async def get_workflow_revisions(
    workflow_id: uuid.UUID,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.get_workflow_revisions(db, workflow_id)


# workflow revisions


@router.get(
    "/workflows/{workflow_id}/revisions/{increment}",
    response_model=schemas.WorkflowRevision,
)
async def get_workflow_revision(
    workflow_id: uuid.UUID,
    increment: int,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.get_workflow_revision(db, workflow_id, increment)


@router.post(
    "/workflows/{workflow_id}/revisions/{increment}/restore",
    response_model=schemas.Workflow,
)
async def restore_workflow_revision(
    workflow_id: uuid.UUID,
    increment: int,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.restore_workflow_revision(db, workflow_id, increment)


# tasks


@router.get("/workflows/{workflow_id}/tasks", response_model=List[schemas.Task])
async def get_workflow_tasks(
    workflow_id: uuid.UUID,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.get_workflow_tasks(db, workflow_id)


@router.get(
    "/workflows/{workflow_id}/tasks/{task_title}",
    response_model=schemas.Task,
)
async def get_task(
    workflow_id: uuid.UUID,
    task_title: str,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.get_task(db, workflow_id, task_title)


@router.get(
    "/workflows/{workflow_id}/revisions/{increment}/tasks",
    response_model=List[schemas.Task],
)
async def get_workflow_revision_tasks(
    workflow_id: uuid.UUID,
    increment: int,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.get_workflow_tasks(db, workflow_id, increment=increment)


@router.get(
    "/workflows/{workflow_id}/revisions/{increment}/tasks/{task_title}",
    response_model=schemas.Task,
)
async def get_revision_task(
    workflow_id: uuid.UUID,
    increment: int,
    task_title: str,
    db: AsyncSession = Depends(get_async_db),
):
    return await service.get_task(db, workflow_id, task_title, increment=increment)

File: services/base/workflow-ui/docker/files/src/api/workflows.ts
import apiClient from './workflowApiClient';
import type {
    Workflow,
    WorkflowCreate,
    WorkflowRevision,
    WorkflowUpdate,
} from '@/types/schemas';


// Fetch all workflows
export async function fetchWorkflows(params?: { title?: string }): Promise<Workflow[]> {
    const response = await apiClient.get<Workflow[]>('/workflows', { params });
    if (!response.data || !Array.isArray(response.data)) {
        return [];
    }
    return response.data;
}

// Fetch a single workflow by UUID
export async function fetchWorkflowById(id: string): Promise<Workflow> {
    const response = await apiClient.get<Workflow>(`/workflows/${id}`);
    return response.data;
}

// Create a new workflow
export async function createWorkflow(workflow: WorkflowCreate): Promise<Workflow> {
    const response = await apiClient.post<Workflow>('/workflows', workflow);
    return response.data;
}

// Patch a workflow (creates a new revision)
export async function patchWorkflow(id: string, update: WorkflowUpdate): Promise<Workflow> {
    const response = await apiClient.patch<Workflow>(`/workflows/${id}`, update);
    return response.data;
}

// Delete a workflow (soft-delete)
export async function deleteWorkflow(id: string): Promise<void> {
    await apiClient.delete(`/workflows/${id}`);
}

// Fetch tasks for a workflow by UUID (current revision)
export async function fetchWorkflowTasks(workflowId: string) {
    const response = await apiClient.get(`/workflows/${workflowId}/tasks`)
    return response.data
}

// List all revisions of a workflow
export async function fetchWorkflowRevisions(workflowId: string): Promise<WorkflowRevision[]> {
    const response = await apiClient.get<WorkflowRevision[]>(`/workflows/${workflowId}/revisions`);
    return response.data;
}

// Fetch a specific revision by increment
export async function fetchWorkflowRevision(
    workflowId: string,
    increment: number,
): Promise<WorkflowRevision> {
    const response = await apiClient.get<WorkflowRevision>(
        `/workflows/${workflowId}/revisions/${increment}`,
    );
    return response.data;
}

// Restore (rewind) a workflow to a specific revision; creates a new increment with that content.
export async function restoreWorkflowRevision(
    workflowId: string,
    increment: number,
): Promise<Workflow> {
    const response = await apiClient.post<Workflow>(
        `/workflows/${workflowId}/revisions/${increment}/restore`,
    );
    return response.data;
}
